cylc/flow/task_events_mgr.py

```diff
@@ -812,7 +814,8 @@ def _retry_task(self, itask, wallclock_time, submit_retry=False):
                 os.getenv("CYLC_SUITE_RUN_DIR")
             )
             itask.state.add_xtrigger(label)
-        itask.state.reset(TASK_STATUS_WAITING)
+        if itask.state.reset(TASK_STATUS_WAITING):
+            self.data_store_mgr.delta_task_state(itask)
```

Comment:
I wrestled with this too..

On the one hand it would be simpler to pass the data_store_mgr object into ever `TaskProxy`, and then `Outputs` and `TaskState`, then to Prerequisites object and so on... You can see it being passed around a lot, like public bicycle but worse..

But on the other hand, the data_store_mgr doesn't need to know about all the task proxies (i.e. runahead and other not ready), and/or all their events...

At least this way we can be intentional about it... but I'm open..

Reply:
Ok, might open a follow-on issue as this isn't entirely a data-store issue. Recently hit a bug where the DB wasn't being updated for a particular state change because the full procedure wasn't being followed, so I'm a bit worried about potential bugs now we have the task proxy, data store and data base to keep in sync.

(Long-term I would really like task events to be pushed to a centralised "event queue" with multiple consumers able to subscribe to the event stream.)

Reply:
For now I guess we could write an callback method in the task_pool and pass that to each task_proxy to avoid having to pass a handful of managers down to every object?

Reply:
Bumped - https://github.com/cylc/cylc-flow/issues/4037